src/Halogen/VirtualDOM/Driver.purs

```diff
@@ -107,166 +39,43 @@ runUI
-  -> HalogenM s f g p o (Aff (HalogenEffects eff))
-  ~> Aff (HalogenEffects eff)
-evalM ref (HalogenM q) = foldFree (eval ref) q
-
-render
-  :: forall s f g eff p o
-   . Ref (LifecycleHandlers eff)
-  -> AVar (DriverState s f g eff p o)
-  -> Aff (HalogenEffects eff) Unit
-render lchs var = takeVar var >>= \(DriverState ds) -> do
-  childrenVar <- liftEff $ newRef M.empty
-  oldChildren <- liftEff $ newRef ds.children
-  let selfEval = evalF ds.selfRef
-  vtree' <-
-    renderHTML
-      (handleAff <<< selfEval)
-      (renderChild selfEval ds.fresh ds.mkOrdBox oldChildren childrenVar lchs)
-      (ds.component.render ds.state)
-  node' <- liftEff $ V.patch (V.diff ds.vtree vtree') ds.node
-  children <- liftEff $ readRef childrenVar
-  liftEff (readRef oldChildren) >>= traverse_ (addFinalizer lchs <=< peekVar)
-  putVar var $
-    DriverState
-      { node: node'
-      , vtree: vtree'
-      , component: ds.component
-      , state: ds.state
-      , children
-      , mkOrdBox: ds.mkOrdBox
-      , selfRef: ds.selfRef
-      , handler: ds.handler
-      , keyId: ds.keyId
-      , fresh: ds.fresh
-      }
-
-renderChild
-  :: forall f g eff p
-   . (f ~> Aff (HalogenEffects eff))
-  -> Ref Int
-  -> (p -> OrdBox p)
-  -> Ref (M.Map (OrdBox p) (AVar (DriverStateX g eff)))
-  -> Ref (M.Map (OrdBox p) (AVar (DriverStateX g eff)))
-  -> Ref (LifecycleHandlers eff)
-  -> ComponentSlot HTML g (Aff (HalogenEffects eff)) p (f Unit)
-  -> Aff (HalogenEffects eff) V.VTree
-renderChild handler fresh mkOrdBox childrenInRef childrenOutRef lchs =
-  unComponentSlot \p ctor k -> do
-    childrenIn <- liftEff $ readRef childrenInRef
-    var <- case M.pop (mkOrdBox p) childrenIn of
-      Just (Tuple existing childrenIn') -> do
-        liftEff $ writeRef childrenInRef childrenIn'
-        pure existing
-      Nothing ->
-        runComponent (maybe (pure unit) handler <<< k) fresh lchs (force ctor)
-    liftEff $ modifyRef childrenOutRef (M.insert (mkOrdBox p) var)
-    pure <<< unDriverStateX (\st -> V.widget st.keyId st.node) =<< peekVar var
-
--- | TODO: we could do something more intelligent now this isn't baked into the
--- | virtual-dom rendering. Perhaps write to an avar when an error occurs...
--- | something other than a runtime exception anyway.
-handleAff
-  :: forall eff a
-   . Aff (HalogenEffects eff) a
-  -> Eff (HalogenEffects eff) Unit
-handleAff = void <<< runAff throwException (const (pure unit))
-
-peekVar :: forall eff a. AVar a -> Aff (avar :: AVAR | eff) a
-peekVar v = do
-  a <- takeVar v
-  putVar v a
-  pure a
+runUI component element = AD.runUI (renderSpec element) component
+
+renderSpec :: forall eff. HTMLElement -> AD.RenderSpec HTML RenderState eff
+renderSpec element =
+  { initRenderState
+  , setup
+  , render
+  , renderChild
+  }
+  where
+
+  initRenderState :: Aff (HalogenEffects eff) RenderState
+  initRenderState =
+    let vtree = V.vtext ""
+    in { vtree, node: _ } <$> liftEff (V.createElement vtree)
```

Comment:
This aspect seems a little strange because you are creating render state before a render has actually been initiated (hence the dummy values), right? If you think other backends will have to do something similar, is this a problem of ordering in component initialization? Is there a way to run the initial render and build the render state off of that?